pkg/controller/node/nodecontroller.go

```diff
@@ -505,3 +526,96 @@ func (nc *NodeController) tryUpdateNodeStatus(node *api.Node) (time.Duration, ap
 
 	return gracePeriod, lastReadyCondition, readyCondition, err
 }
+
+// returns true if the provided node still has pods scheduled to it, or an error if
+// the server could not be contacted.
+func (nc *NodeController) hasPods(nodeID string) (bool, error) {
+	pods, err := nc.kubeClient.Pods(api.NamespaceAll).List(labels.Everything(), fields.OneTermEqualSelector(client.PodHost, nodeID))
+	if err != nil {
+		return false, err
+	}
+	return len(pods.Items) > 0, nil
+}
+
+// deletePods will delete all pods from master running on given node, and return true
+// if any pods were deleted.
+func (nc *NodeController) deletePods(nodeID string) (bool, error) {
+	remaining := false
+	pods, err := nc.kubeClient.Pods(api.NamespaceAll).List(labels.Everything(), fields.OneTermEqualSelector(client.PodHost, nodeID))
+	if err != nil {
+		return remaining, err
+	}
+
+	if len(pods.Items) > 0 {
+		nc.recordNodeEvent(nodeID, "DeletingAllPods", fmt.Sprintf("Deleting all Pods from Node %v.", nodeID))
+	}
+
+	for _, pod := range pods.Items {
+		// Defensive check, also needed for tests.
+		if pod.Spec.NodeName != nodeID {
+			continue
+		}
+		// if the pod has already been deleted, ignore it
+		if pod.DeletionGracePeriodSeconds != nil {
```

Comment:
It's defaulted by the API so cannot be nil - it becomes 30s everywhere

On Aug 24, 2015, at 10:33 AM, Marek Grabowski <notifications@github.com>
wrote:

In pkg/controller/node/nodecontroller.go
<https://github.com/kubernetes/kubernetes/pull/12962#discussion_r37750787>:

> +	pods, err := nc.kubeClient.Pods(api.NamespaceAll).List(labels.Everything(), fields.OneTermEqualSelector(client.PodHost, nodeID))
> +	if err != nil {
> +		return remaining, err
> +	}
> +
> +	if len(pods.Items) > 0 {
> +		nc.recordNodeEvent(nodeID, "DeletingAllPods", fmt.Sprintf("Deleting all Pods from Node %v.", nodeID))
> +	}
> +
> +	for _, pod := range pods.Items {
> +		// Defensive check, also needed for tests.
> +		if pod.Spec.NodeName != nodeID {
> +			continue
> +		}
> +		// if the pod has already been deleted, ignore it
> +		if pod.DeletionGracePeriodSeconds != nil {

Is this backward compatible? IIUC DeletionGracePeriodSeconds was introduced
quite recently, and thus will be nil in all 'upgraded' clusters.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12962/files#r37750787>.